src/datapoints-graph/charts/echarts-options.service.ts

```diff
@@ -111,24 +111,242 @@ export class EchartsOptionsService {
         },
       },
       yAxis,
-      series: this.getChartSeries(datapointsWithValues),
+      series: this.getChartSeries(datapointsWithValues, events, alarms),
     };
   }
 
   private getChartSeries(
-    datapointsWithValues: DatapointWithValues[]
+    datapointsWithValues: DatapointWithValues[],
+    events,
+    alarms
   ): SeriesOption[] {
     const series: SeriesOption[] = [];
+    let eventSeries: SeriesOption[] = [];
+    let alarmSeries: SeriesOption[] = [];
     datapointsWithValues.forEach((dp, idx) => {
       const renderType: DatapointChartRenderType = dp.renderType || 'min';
       if (renderType === 'area') {
         series.push(this.getSingleSeries(dp, 'min', idx, true));
         series.push(this.getSingleSeries(dp, 'max', idx, true));
       } else {
-        series.push(this.getSingleSeries(dp, renderType, idx));
+        series.push(this.getSingleSeries(dp, renderType, idx, false));
       }
+
+      const newEventSeries = this.getAlarmOrEventSeries(
+        dp,
+        renderType,
+        false,
+        events,
+        'event'
+      );
+      const newAlarmSeries = this.getAlarmOrEventSeries(
+        dp,
+        renderType,
+        false,
+        alarms,
+        'alarm'
+      );
+      eventSeries = [...eventSeries, ...newEventSeries];
+      alarmSeries = [...alarmSeries, ...newAlarmSeries];
     });
-    return series;
+    return [...series, ...eventSeries, ...alarmSeries];
+  }
+
+  groupByType(items: any[], typeField: string, hiddenField: string) {
```

Comment:
lack of items type (we can tell what can be type of items, no need to use any), also lack of return type

Reply:
it's actually grouping by type AND filter out hidden items. this method is used only in one place, so I suggest to make it simple and filter items list first, then call groupByType